File: python/src/keycloak_sdk/tokens.py
"""토큰 값 타입."""

from __future__ import annotations

import math
from collections.abc import Mapping
from dataclasses import dataclass, field
from types import MappingProxyType
from typing import Any

from ._internal.secrets import mask
from .exceptions import KeycloakAuthError


@dataclass(frozen=True)
class TokenSet:
    access_token: str
    refresh_token: str | None
    id_token: str | None
    token_type: str
    scope: str | None
    expires_at: float | None

    def is_expired(self, now: float, skew: float) -> bool:
        if self.expires_at is None:
            return True
        return now + skew >= self.expires_at

    def __repr__(self) -> str:
        return (
            f"TokenSet(token_type={self.token_type!r}, scope={self.scope!r}, "
            f"access_token={mask(self.access_token)!r}, "
            f"refresh_token={mask(self.refresh_token)!r}, expires_at={self.expires_at!r})"
        )

    @staticmethod
    def from_response(data: dict[str, Any], issued_at: float) -> TokenSet:
        # ⚠️ **존재 검사는 타입 검사가 아니다.** 예전에는 `data["access_token"]` 이라
        # 숫자·객체가 그대로 들어왔고(타입 힌트는 `str`), 소비자는 그것을 Bearer 로 실어
        # 보내 매번 401 을 받았다. 키가 없으면 raw `KeyError` 가 새서 `keycloak_sdk.
        # exceptions` 를 잡는 소비자가 **아무것도 잡지 못했다**(§4). 나머지 필드도 같다 —
        # `expires_in` 이 숫자가 아니면 raw `ValueError`/`TypeError` 가 샜고, 앞의 것은 **값을
        # 인용한다**. 객체 모양 `token_type` 은 `repr` 이 그대로 찍었다(실측 2026-09-26).
        # ⚠️ 거부 메시지에는 필드 이름만 싣는다 — 형식이 틀린 응답은 그 자리에 토큰을 실어 온다.
        access_token = data.get("access_token")
        if not isinstance(access_token, str) or not access_token:
            raise KeycloakAuthError("token response has no usable access_token")
        return TokenSet(
            access_token=access_token,
            refresh_token=_optional_str(data, "refresh_token"),
            id_token=_optional_str(data, "id_token"),
            token_type=_optional_str(data, "token_type") or "Bearer",
            scope=_optional_str(data, "scope"),
            expires_at=_expires_at(data.get("expires_in"), issued_at),
        )


def _optional_str(data: dict[str, Any], name: str, what: str = "token response") -> str | None:
    value = data.get(name)
    if value is None or isinstance(value, str):
        return value
    raise KeycloakAuthError(f"{what} has an invalid {name}")


def _expires_at(expires_in: Any, issued_at: float) -> float | None:
    if expires_in is None:
        return None
    try:
        seconds = float(expires_in)  # 숫자 문자열("300")도 받는다 — 예전과 같다
    except (TypeError, ValueError):
        seconds = math.nan
    # ⚠️ 거부는 `except` **밖**이다 — 안에서 던지면 값을 인용한 `ValueError` 가 `__context__` 로
    # 매달려 `logging.exception` 이 찍는다. 무한대·NaN 도 거부한다(`is_expired` 가 영원히 거짓).
    if not math.isfinite(seconds):
        raise KeycloakAuthError("token response has an invalid expires_in")
    return issued_at + seconds


@dataclass(frozen=True)
class ValidatedToken:
    subject: str | None
    issuer: str | None
    audience: tuple[str, ...]
    expires_at: float | None
    issued_at: float | None
    claims: Mapping[str, Any] = field(default_factory=dict)

    def __post_init__(self) -> None:
        # frozen dataclass가 claims 필드 자체 재바인딩은 막지만 dict 내부 변이는 막지
        # 못한다 — 읽기전용 MappingProxyType으로 감싸 소비자의 내부 변이를 차단한다
        # (Java ValidatedToken.getClaims()의 unmodifiableMap과 동형).
        object.__setattr__(self, "claims", MappingProxyType(dict(self.claims)))


@dataclass(frozen=True)
class IntrospectionResult:
    active: bool
    username: str | None
    client_id: str | None


def _introspection_result(data: dict[str, Any]) -> IntrospectionResult:
    """introspect 응답 → `IntrospectionResult`. sync·aio 미러가 같이 쓴다(RFC 7662).

    ⚠️ `active` 는 JSON boolean 만 받는다 — 예전 `bool(...)` 은 문자열 `"false"` 를 **활성**으로
    읽었다(fail-open, 실측 2026-09-26). 없거나 null 이면 비활성이다(예전과 같다). `username`·
    `client_id` 는 str|None 이다 — 객체 모양은 기본 dataclass repr 이 그 안의 토큰을 찍었다(Grok
    레그가 찾음). 거부 메시지에는 필드 이름만 싣는다."""
    what = "introspection response"
    active = data.get("active")
    if active is None:
        active = False
    if not isinstance(active, bool):
        raise KeycloakAuthError(f"{what} has an invalid active")
    return IntrospectionResult(
        active=active,
        username=_optional_str(data, "username", what),
        client_id=_optional_str(data, "client_id", what),
    )

File: ruby/lib/keycloak_sdk/tokens.rb
# frozen_string_literal: true

module KeycloakSdk
  # OAuth 토큰 응답의 불변 값타입. access/refresh/id 토큰은 inspect에서 마스킹.
  TokenSet = Data.define(:access_token, :token_type, :expires_in, :refresh_token,
                         :id_token, :scope, :expires_at) do
    # ⚠️ **검증은 팩토리가 아니라 생성자에 있다.** `from_response` 에만 두면 `AuthClient`
    # 가 `TokenSet.new` 을 직접 부르는 경로(`auth_client.rb` 의 `to_token_set`)가 그것을
    # 통째로 우회한다 — 독립 레그가 그 구멍을 지목했다. 생성자는 어떤 경로도 지나므로
    # 여기 두면 우회가 불가능하다.
    def initialize(access_token:, **rest)
      unless access_token.is_a?(String) && !access_token.empty?
        raise AuthError, "token response has no usable access_token"
      end

      super
    end

    def self.from_response(body, received_at: Time.now.to_f)
      # ⚠️ **존재 검사는 타입 검사가 아니다.** 예전에는 값을 그대로 담아 숫자·해시·nil 이
      # `access_token` 이 됐고, 소비자는 그것을 Bearer 로 실어 보내 매번 401 을 받았다
      # (조용한 반복 실패). `expires_in` 의 문자열 허용은 **의도된 것**이라 건드리지 않는다.
      access_token = body["access_token"]
      unless access_token.is_a?(String) && !access_token.empty?
        raise AuthError, "token response has no usable access_token"
      end

      expires_in = body["expires_in"] && Integer(body["expires_in"])
      new(
        access_token: access_token,
        token_type: body["token_type"],
        expires_in: expires_in,
        refresh_token: body["refresh_token"],
        id_token: body["id_token"],
        scope: body["scope"],
        expires_at: expires_in ? received_at + expires_in : nil
      )
    end

    # ⚠️ **만료 시각을 모르면 "만료됨"이다**(fail-safe — 자매 여덟과 동형, Java 의 M.6).
    # `false`(=아직 살아있다)를 돌려주면 만료 시각 미상인 토큰이 **영원히 유효**해진다.
    # `expires_at` 이 nil 인 경우는 서버가 `expires_in` 을 안 보냈을 때뿐이라 정상 경로가
    # 아니고, 그때 취할 안전한 쪽은 "재발급"이다.
    def expired?(skew: 0, now: Time.now.to_f)
      return true if expires_at.nil?

      now >= (expires_at - skew)
    end

    def inspect
      "#<KeycloakSdk::TokenSet access_token=\"***\" token_type=#{token_type.inspect} " \
        "expires_in=#{expires_in.inspect} refresh_token=#{refresh_token ? '"***"' : 'nil'} " \
        "id_token=#{id_token ? '"***"' : 'nil'} scope=#{scope.inspect} expires_at=#{expires_at.inspect}>"
    end
    alias_method :to_s, :inspect

    # ⚠️ `pp`/`pretty_inspect` 는 `inspect` 와 같은 계급의 **표시 경로**인데, `Data` 타입은 PP 가
    # 멤버를 직접 찍어 위 `inspect` 를 **건너뛴다**(실측: access_token 원문이 나왔다). 일반
    # 클래스인 `Config` 는 PP 가 `inspect` 로 폴백해 안전하다 — 그래서 `Data` 만 이 훅이 필요하다.
    def pretty_print(pp)
      pp.text(inspect)
    end
  end

  # 검증된 access token의 관심 클레임.
  ValidatedToken = Data.define(:subject, :audience, :issuer, :expires_at, :issued_at, :claims)

  # RFC 7662 introspection 결과.
  IntrospectionResult = Data.define(:active, :username, :client_id, :claims) do
    def self.from_response(body)
      new(active: body["active"] == true, username: body["username"],
          client_id: body["client_id"], claims: body)
    end

    def active?
      active == true
    end
  end

  # authorization-code 흐름 시작 값(PKCE code_verifier 포함·inspect 마스킹).
  # nonce는 인가 URL에 실리는 재생 방지 값이라 비밀이 아니다(state와 동급 — code_verifier만 마스킹).
  AuthorizationRequest = Data.define(:url, :state, :code_verifier, :nonce) do
    def inspect
      "#<KeycloakSdk::AuthorizationRequest url=#{url.inspect} state=#{state.inspect} " \
        "nonce=#{nonce.inspect} code_verifier=\"***\">"
    end
    alias_method :to_s, :inspect

    # ⚠️ `pp`/`pretty_inspect` 는 `inspect` 와 같은 계급의 **표시 경로**인데, `Data` 타입은 PP 가
    # 멤버를 직접 찍어 위 `inspect` 를 **건너뛴다**(실측: access_token 원문이 나왔다). 일반
    # 클래스인 `Config` 는 PP 가 `inspect` 로 폴백해 안전하다 — 그래서 `Data` 만 이 훅이 필요하다.
    def pretty_print(pp)
      pp.text(inspect)
    end
  end
end
